### PR TITLE
BIG-4192 - Use `rep-cpp`

### PR DESCRIPTION
**Pull request by dlecocq**
This uses `rep-cpp` to replace the core functionality here.

On `master`:

```
(venv) vagrant@reppy:/vagrant$ ./bench.py 
Parse
==========
Total: 3.57568597794
  Avg: 3.57568597794e-05
 Rate: 27966.6616747

Evaluate
==========
Total: 0.284116029739
  Avg: 2.84116029739e-06
 Rate: 351968.877264
```

On this branch:

```
(venv) vagrant@reppy:/vagrant$ ./bench.py 
Parse
==========
Total: 0.708585977554
  Avg: 7.08585977554e-06
 Rate: 141126.134538

Evaluate
==========
Total: 0.0823559761047
  Avg: 8.23559761047e-07
 Rate: 1214240.96623
```

That's roughly a 5x parse improvement and about 3.5x evaluation improvement.

@neilmb @b4hand @lindseyreno

### CODE REVIEW

File: tests/test_agent.py [hunk 1]
```diff
@@ -1,63 +1,90 @@
...
         '''Handles the case where the wildcard rule is unescaped.'''
-        agent = Agent().disallow('/a<*')
+        agent = self.parse('''
+            User-agent: agent
+            Disallow: /a<*
+        ''', 'agent')
         self.assertFalse(agent.allowed('/a<d.html'))
         self.assertFalse(agent.allowed('/a%3cd.html'))
 
     def test_allow_url(self):
```
Comment (b4hand) [inline tests/test_agent.py]: I would change the name of this case as well. I mentioned it in the other PR.

**Comment by kmike** [comment]
Does it mean there won't be a pure-Python version of robots.txt parsing?

**Comment by dlecocq** [comment]
@kmike If it's important to the community, we could preserve that. Are you using pypi?

**Comment by kmike** [comment]
@dlecocq we're considering reppy as a replacement for stdlib robotparser for Scrapy because it supports much more robots.txt files from the real world, and because it handles Crawl-Delay (which is only implemented in most recent Python 3). It is also easier to get fixes to a non-stdlib package.

It is more convenient for users to have pure-Python packages with optional speedups. Installing Scrapy is hard enough already because of lxml/cffi/cryptography packages.

There are some other points we're unsure of:

* it seems the repo in under a heavy refactoring now, so it looks like we should wait before making a decision and writing code;
* currently reppy has both robots.txt parsing and logic for TTL and for fetching robots.txt files from remote websites; it is separated well, but we're only interested in robots.txt parsing - making sync requests from a library is a no-go for us. It'd also be weird to have `requests` library as a Scrapy dependency, but it is in reppy install_requires.

**Comment by dlecocq** [comment]
@kmike I totally agree about it being easier to deal with pure python. Even if we kept a pure python, we depend on https://github.com/seomoz/url-py, which also uses a C++ extension.

In both cases, we support builds with and without Cython, so at least there's that. I imagine most instances have a recent-ish compiler (though both `url-cpp` and `rep-cpp` require C++11.

We could turn `requests` into an optional dependency. One of the things we really made an effort to do in this refactoring has been to better separate the fetching from the parsing in order to support cases like the one you describe.

In terms of the performance boost, it's about 5x for parsing, about 3.5x for evaluating. I'm not sure how much time your benchmarks show you're currently using for most Scrapy applications, but for us it ends up being pretty heavy for some of our applications -- especially for extremely FQDN-diverse request streams.

**Comment by kmike** [comment]
> @kmike I totally agree about it being easier to deal with pure python. Even if we kept a pure python, we depend on https://github.com/seomoz/url-py, which also uses a C++ extension.

Ah, I didn't know that. We were also looking at URL parsing speed performance, as it can be a bottleneck in spiders (stdlib parsing is slow) - https://github.com/scrapy/scrapy/issues/1306. So we're also looking for URL parsing library :) Ideally it should be fast and compatible with https://url.spec.whatwg.org/, or just URL parsing code extracted from a real browser. 

By the way, it seems that url-py doesn implement url.escape correctly: in URLs path should be always escaped to utf8 and query arguments should be escaped to web page encoding; this is crazy, but this is how all browsers work with non-ascii urls.

> In both cases, we support builds with and without Cython, so at least there's that. I imagine most instances have a recent-ish compiler (though both url-cpp and rep-cpp require C++11.

It is complicated for us - Scrapy userbase is very diverse, there is a lot of outdated Linux machines, Windows users (up to 50% according to website stats) and non-Python developers. cryptography/cffi/lxml usually provide wheels for different OSes, but we're still getting a stream of bug reports and support questions about installing Scrapy.

> We could turn requests into an optional dependency. One of the things we really made an effort to do in this refactoring has been to better separate the fetching from the parsing in order to support cases like the one you describe.

Sounds good!

> In terms of the performance boost, it's about 5x for parsing, about 3.5x for evaluating. I'm not sure how much time your benchmarks show you're currently using for most Scrapy applications, but for us it ends up being pretty heavy for some of our applications -- especially for extremely FQDN-diverse request streams.

URL parsing can be a bottleneck, but I'm not sure about robots.txt parsing because the file is only parsed once per site during a crawl; hm, it could be a problem for some of the broad crawls.

**Comment by dlecocq** [comment]
If URL parsing is slowing you down, `url-py` is substantially faster than python 2.7's implementation, especially once you start doing normalization tasks (path normalization, relativizing, absolutizing, escaping, etc.). For our workloads, it's around 20-50x faster than when we were using the standard libs. For our crawling efforts, this gave us roughly a 4x net gain in crawling and parsing.

Re: escape, that's... huh. When you say the web page encoding, you mean in the context of the web page on which it was found? Could you point me to an RFC or the like that describes that?

The switch for us materialized as for many-to-most aspects, a direct port of python to C++ of the standard library parsing, if for nothing else, to ease transition. There were some exceptions to that rule, of course.

One thing that I'll note is lacking in `url-py` is to do with the normalization of codepoints for non-ascii hostnames. It does handle the punycoding, _except_ for the codepoint normalization step -- it should be a matter of copying a bunch of codepoint tables from the relevant RFC, or adding an external dependency (which we've been very reluctant to do).

**Comment by kmike** [comment]
> If URL parsing is slowing you down, url-py is substantially faster than python 2.7's implementation, especially once you start doing normalization tasks (path normalization, relativizing, absolutizing, escaping, etc.). For our workloads, it's around 20-50x faster than when we were using the standard libs. For our crawling efforts, this gave us roughly a 4x net gain in crawling and parsing.

Yeah, a good point; just a faster urlparse is not going to give this speedup (though it still get a speedup).

> Re: escape, that's... huh. When you say the web page encoding, you mean in the context of the web page on which it was found? Could you point me to an RFC or the like that describes that?

Yes, encoding of a web page URL was found. It is not a part of RFC, but it is in https://url.spec.whatwg.org/ - it seems this resource formalizes what all browsers do anyways. We discovered it by creating test pages and trying several browsers; I wrote comment on [HN](https://news.ycombinator.com/item?id=11674220) about that, and got some awesome comments which explain why did it happen.

> The switch for us materialized as for many-to-most aspects, a direct port of python to C++ of the standard library parsing, if for nothing else, to ease transition. There were some exceptions to that rule, of course.

It looks like https://github.com/commonsearch/urlparse4 also provides urllib-like interface; it has some benchmarks. It'd be nice to add url-py there, even though speeding up urllib functions is not the main benefit (normalization, etc. speedups can be larger).

> One thing that I'll note is lacking in url-py is to do with the normalization of codepoints for non-ascii hostnames. It does handle the punycoding, except for the codepoint normalization step -- it should be a matter of copying a bunch of codepoint tables from the relevant RFC, or adding an external dependency (which we've been very reluctant to do).

Hm, do you know if hostname.encode('idna') is enough? Are there examples where this is not the way to go? encode('idna') is what we're using in w3lib.url (and thus in scrapy).

**Comment by dlecocq** [comment]
It could be pretty annoying to handle other encodings, but in the worst case, non-utf8 encodings could fall back to a slower (read: python) implementation.

I gave `urlparse4` a quick test locally, and for the examples I tried, `url-py` edges it out by about 20-25% for just parsing, so not a huge difference.

Yes, I believe `hostname.encode('idna')` is sufficient. Punycoding is not a path we've spent much time optimizing (though we did port it to C++, save the codepoint normalization) mostly because for our purposes it's in the vast, vast minority. As such, we're not opposed to taking perf hits in that cast for correctness.

**Comment by b4hand** [comment]
@kmike: Regarding https://url.spec.whatwg.org/ and url-py/url-cpp, we're specifically aware of that standard. There are some places where url-py/url-cpp differ from that standard mainly because the original normalization code we've used predates that standard. We've talked about getting a closely conforming implementation, but it is not fully there yet. We also heavily rely on this code internally for a variety of projects, so we're not always able to make dramatic changes in the results quickly, but in the long run we do plan to make it as conformant as possible.

**Comment by dlecocq** [comment]
Reupped. I added a README, and it's now using `rep-cpp@master`

**Comment by dlecocq** [comment]
I expect this build to now pass on 3.3+ as well.

**Pull request merged**